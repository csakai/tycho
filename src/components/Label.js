import React from 'react';
import PropTypes from 'prop-types';

class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    text: PropTypes.string.isRequired,
    active: PropTypes.bool
  }

  getPosition = (pos) => {
    if (pos) {
      return {
        position: 'absolute',
        top: `${pos.top}px`,
        left: `${pos.left}px`
      };
    }
    return null;
  }
  
  render() {
    return (
      <div>
        <span style={this.getPosition(this.props.position)}>
          {this.props.text}
        </span>
        {this.props.active ? this.props.children : null}
      </div>
    );
  }
}

export default Label;
