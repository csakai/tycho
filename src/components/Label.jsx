import React from 'react';
import PropTypes from 'prop-types';

class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    text: PropTypes.string.isRequired,
    active: PropTypes.bool
  }

  getPosition = (pos) => {
    if (pos) {
      return {
        position: 'absolute',
        top: `${pos.top}px`,
        left: `${pos.left}px`
      };
    }
    return null;
  }

  getChildren = (children, active) => {
    if (active) {
      return children;
    }
    return null;
  }
  
  render() {
    return (
      <div>
        <span style={this.getPosition(this.props.position)}>
          {this.props.text}
        </span>
        {this.getChildren(this.props.children, this.props.active)}
      </div>
    );
  }
}

export default Label;
