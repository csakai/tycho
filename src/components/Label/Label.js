import React from 'react';
import PropTypes from 'prop-types';

class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    orbital: PropTypes.object.isRequired
  }

  getPosition = () => {
    const pos = this.props.position;

    if (pos) {
      return {
        position: 'absolute',
        top: `${pos.top}px`,
        left: `${pos.left}px`
      };
    }
    return null;
  }
  
  render() {
    return (
      <span style={this.getPosition()}>
        {this.props.orbital.name}
      </span>
    );
  }
}

export default Label;
