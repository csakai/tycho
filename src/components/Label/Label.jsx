import React from 'react';
import PropTypes from 'prop-types';

export default class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    text: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
    active: PropTypes.bool,
    enabled: PropTypes.bool
  }
  
  render() {
    return (
      <div className="label" style={{
        display: this.props.enabled ? 'block' : 'none'
      }}>
        <span
          className="label__text"
          style={this.props.position}
          onClick={this.props.onClick}>
          {!this.props.active && this.props.text}
        </span>
        {this.props.active && this.props.children}
      </div>
    );
  }
}
