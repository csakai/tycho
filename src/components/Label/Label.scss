.label {
  &__text {
    position: absolute;
    text-shadow: 1px 1px 1px #c0c0c0;
    opacity: 0.75;
    font-family: Georgia, serif;
    cursor: pointer;

    &:hover {
      opacity: 1;
    }
  }
}
