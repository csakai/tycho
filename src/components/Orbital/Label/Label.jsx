import React from 'react';
import PropTypes from 'prop-types';
import DomLabel from 'three-dom-label';

export default class Label extends React.Component {

    static propTypes = {
        text: PropTypes.string.isRequired,
        id: PropTypes.string.isRequired,
        action: PropTypes.object.isRequired,
        targetId: PropTypes.string
    }

    componentDidMount = () => {
        this.label = this.getLabel();   
        this.refs.group.add(this.label);
    }

    componentWillUnmount = () => this.label.unmount()

    componentWillReceiveProps = (nextProps) => {
        this.maybeUpdateClassName(nextProps);
    }

    /**
     * Updates class name if the active target has changed.
     * 
     * @param {Object} props
     * @param {String} props.id - id of current orbital
     * @param {String} props.targetId - id of active orbital
     */
    maybeUpdateClassName = ({id, targetId}) => {
        if (this.props.targetId !== targetId) {
            const isActive = id === targetId;
            const classSuffix = isActive ? 'active' : 'inactive';

            this.label.setClass(`label label--${classSuffix}`);
        }
    }

    /**
     * Renders a new label, with mouse events bound to it.
     * 
     * @returns {Object3D} label
     */
    getLabel = () => {
        const {id, text, action} = this.props;
        const classPrefix = `label label`;

        return new DomLabel({
            ...this.props,
            events: {
                click: action.setActiveOrbital.bind(this, id, text),
                mouseover: action.addHighlightedOrbital.bind(this, id),
                mouseout: action.removeHighlightedOrbital.bind(this, id)
            },
            className: `${classPrefix}--inactive`
        });
    }

    render() {
        return <group ref="group"></group>;
    }
}