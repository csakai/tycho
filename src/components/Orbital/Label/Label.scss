@import 'globals.scss';

.label {
    &--active {
        animation: glowing-text $transition-speed infinite;
        pointer-events: none;
        font-size: $font-size-medium;
    }

    &--inactive {
        color: $brand-primary;
        cursor: pointer;
        transition: font-size $transition-speed;

        &:hover {
            text-shadow: $text-shadow;
            font-size: $font-size-medium;
        }
    }
}

@keyframes glowing-text {
  0%   { text-shadow: 0 }
  40%  { text-shadow: $text-shadow }
  60%  { text-shadow: $text-shadow }
  100% { text-shadow: 0 }
}