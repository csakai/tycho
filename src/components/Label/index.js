import Label from './Label';
import './Label.css';

export default Label;
