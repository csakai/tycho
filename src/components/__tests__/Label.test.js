import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import Label from '../Label';

describe('Label Component', () => {
  let component, label;

  beforeEach(() => {
    component = shallow(<Label 
      text="foo"
      positions={{}}
    />);

    label = component.instance();
  });

  describe('getPosition()', () => {
    it('should return a style object with top, left, and position definitions', () => {
      const pos = {top: 4, left: 10};
      const style = label.getPosition(pos);

      expect(typeof style).toBe('object');
      expect(style).toHaveProperty('position');
      expect(style).toHaveProperty('top');
      expect(style).toHaveProperty('left');
      expect(style.position).toEqual('absolute');
      expect(style.top).toEqual(`${pos.top}px`);
      expect(style.left).toEqual(`${pos.left}px`);
    });

    it('should return null if the position parameter is not defined', () => {
      const style = label.getPosition();

      expect(style).toBeDefined();
      expect(style).toBeNull();
    });
  });

  describe('render()', () => {
    it('should render the Label successfully', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
