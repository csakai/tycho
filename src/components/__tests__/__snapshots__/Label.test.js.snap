// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Component render() should render the Label successfully 1`] = `
<div>
  <span
    style={null}
  >
    foo
  </span>
</div>
`;
